File: filesim.py
#!/usr/bin/env python3
"""
filesim: Recursive file similarity analysis using hybrid NCD + entropy distance.

Concurrency model
-----------------
zstd releases the GIL during compression, so ThreadPoolExecutor gives true
parallelism for CPU-bound compression work with near-zero inter-thread
serialization overhead.  ProcessPoolExecutor was benchmarked and found to be
~4x *slower* on this workload because pickling file bytes across process
boundaries costs more than the GIL savings.

Memory model
------------
After the per-file preprocessing pass we hold only derived metadata
(~19x smaller than raw bytes for typical files):
  - cx              : compressed size (int)
  - H               : global Shannon entropy (float)
  - profile         : per-block entropy list (list[float])
  - fp              : entropy fingerprint serialized to bytes (bytes)
  - dict_bytes      : serialized zstd dictionary (bytes | None)

Raw file bytes are released immediately after preprocessing each file.
Pair computation re-reads files from disk on demand — for small datasets
the OS page cache makes this essentially free; for large datasets it trades
memory for I/O, which is the correct tradeoff when N is large.

Pair pipeline
-------------
itertools.combinations() is a lazy generator — the full O(N^2) pair list is
never materialized in memory.  Pairs are submitted to the thread pool as a
stream; results flow out via as_completed() so fast pairs don't wait on slow
ones.  CSV mode streams directly to stdout without buffering.  Table mode
buffers into a list only for the final sort step.
"""

import argparse
import csv
import math
import os
import struct
import sys
from collections import Counter
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass, field
from itertools import combinations
from pathlib import Path
from typing import Generator, Iterator

import zstandard as zstd


# ---------------------------------------------------------------------------
# Tuning constants
# ---------------------------------------------------------------------------

ZSTD_LEVEL = 10
MIN_DICT_SAMPLES = 5
MIN_SAMPLE_SIZE = 8
ENTROPY_BLOCK_SIZE = 256
DICT_SIZE = 4096
DEFAULT_WORKERS = max(1, (os.cpu_count() or 2) - 1)


# ---------------------------------------------------------------------------
# Per-file metadata (replaces holding raw bytes after preprocessing)
# ---------------------------------------------------------------------------

@dataclass(slots=True)
class FileMeta:
    path: Path
    size: int
    cx: int            # compressed size, no dictionary
    H: float           # global Shannon entropy [0, 8]
    profile: list      # per-block entropy values
    fp: bytes          # entropy fingerprint (packed floats)
    dict_bytes: bytes | None  # serialized zstd dict


# ---------------------------------------------------------------------------
# Entropy utilities
# ---------------------------------------------------------------------------

def _shannon_entropy(data: bytes) -> float:
    if not data:
        return 0.0
    counts = Counter(data)
    n = len(data)
    return -sum((c / n) * math.log2(c / n) for c in counts.values())


def _entropy_profile(data: bytes) -> list:
    result = []
    for i in range(0, len(data), ENTROPY_BLOCK_SIZE):
        blk = data[i : i + ENTROPY_BLOCK_SIZE]
        if len(blk) >= MIN_SAMPLE_SIZE:
            result.append(_shannon_entropy(blk))
    return result or [_shannon_entropy(data)]


def _entropy_fingerprint(profile: list) -> bytes:
    return struct.pack(f"{len(profile)}f", *profile)


def _profile_distance(p1: list, p2: list) -> float:
    n = min(len(p1), len(p2))
    if n == 0:
        return 1.0
    return sum(abs(a - b) for a, b in zip(p1[:n], p2[:n])) / (n * 8.0)


# ---------------------------------------------------------------------------
# Compression utilities
# ---------------------------------------------------------------------------

def _compress_size(data: bytes, dictionary=None) -> int:
    """Compress data and return byte length.  zstd releases the GIL here."""
    c = zstd.ZstdCompressor(level=ZSTD_LEVEL, dict_data=dictionary)
    return len(c.compress(data))


def _build_dict_bytes(data: bytes) -> bytes | None:
    """Train a zstd dict, return raw bytes (serializable across threads)."""
    chunk_size = max(len(data) // 10, MIN_SAMPLE_SIZE)
    samples = [
        data[i : i + chunk_size]
        for i in range(0, len(data), chunk_size)
        if len(data[i : i + chunk_size]) >= MIN_SAMPLE_SIZE
    ]
    if len(samples) < MIN_DICT_SAMPLES:
        return None
    try:
        d = zstd.train_dictionary(DICT_SIZE, samples)
        return d.as_bytes()
    except Exception:
        return None


def _dict_from_bytes(raw: bytes | None):
    if raw is None:
        return None
    d = zstd.ZstdCompressionDict(raw)
    d.precompute_compress(level=ZSTD_LEVEL)
    return d


# ---------------------------------------------------------------------------
# Preprocessing: file -> FileMeta  (raw bytes released on return)
# ---------------------------------------------------------------------------

def _preprocess_file(path: Path) -> FileMeta:
    data = path.read_bytes()
    meta = FileMeta(
        path=path,
        size=len(data),
        cx=_compress_size(data),
        H=_shannon_entropy(data),
        profile=_entropy_profile(data),
        fp=_entropy_fingerprint(_entropy_profile(data)),
        dict_bytes=_build_dict_bytes(data),
    )
    return meta  # data goes out of scope here


# ---------------------------------------------------------------------------
# Pair computation (unit of work for thread pool)
# ---------------------------------------------------------------------------

def _ncd_fp(fp_a: bytes, fp_b: bytes) -> float:
    ca = _compress_size(fp_a)
    cb = _compress_size(fp_b)
    cab = _compress_size(fp_a + fp_b)
    denom = max(ca, cb)
    return ((cab - min(ca, cb)) / denom) if denom else 0.0


def _compute_pair(a: FileMeta, b: FileMeta, alpha: float, beta: float) -> dict:
    """
    Compute hybrid similarity for one pair.
    Re-reads file bytes on demand; ZstdCompressionDict objects are
    constructed locally and never shared between threads.
    """
    dict_a = _dict_from_bytes(a.dict_bytes)
    dict_b = _dict_from_bytes(b.dict_bytes)

    data_a = a.path.read_bytes()
    data_b = b.path.read_bytes()

    if dict_a or dict_b:
        c_b_given_a = _compress_size(data_b, dict_a) if dict_a else _compress_size(data_b)
        c_a_given_b = _compress_size(data_a, dict_b) if dict_b else _compress_size(data_a)
        denom = max(a.cx, b.cx)
        nd = (min(c_b_given_a, c_a_given_b) / denom) if denom else 0.0
    else:
        cxy = _compress_size(data_a + data_b)
        denom = max(a.cx, b.cx)
        nd = ((cxy - min(a.cx, b.cx)) / denom) if denom else 0.0
    nd = max(0.0, min(1.0, nd))

    del data_a, data_b

    nf = max(0.0, min(1.0, _ncd_fp(a.fp, b.fp)))
    eg = abs(a.H - b.H) / 8.0
    ep = _profile_distance(a.profile, b.profile)

    gamma = (1.0 - alpha - beta) / 2.0
    hybrid = alpha * nd + beta * nf + gamma * eg + gamma * ep

    return {
        "file_a": str(a.path),
        "file_b": str(b.path),
        "hybrid": round(hybrid, 6),
        "ncd_dict": round(nd, 6),
        "ncd_fingerprint": round(nf, 6),
        "entropy_global": round(eg, 6),
        "entropy_profile": round(ep, 6),
        "H_a": round(a.H, 3),
        "H_b": round(b.H, 3),
    }


# ---------------------------------------------------------------------------
# Pipeline
# ---------------------------------------------------------------------------

def preprocess_files(files: list, workers: int, verbose: bool) -> list:
    metas = []
    with ThreadPoolExecutor(max_workers=workers) as pool:
        future_map = {pool.submit(_preprocess_file, f): f for f in files}
        for future in as_completed(future_map):
            path = future_map[future]
            try:
                meta = future.result()
                metas.append(meta)
                if verbose:
                    print(
                        f"  preprocessed {meta.path.name}"
                        f"  ({meta.size:,}B  cx={meta.cx:,}B  H={meta.H:.2f})",
                        file=sys.stderr,
                    )
            except OSError as e:
                print(f"  Warning: cannot read {path}: {e}", file=sys.stderr)
    return metas


def pair_results(
    metas: list,
    alpha: float,
    beta: float,
    workers: int,
    verbose: bool,
) -> Generator:
    """
    Lazy generator yielding result dicts as pairs complete.

    Uses combinations() as a lazy iterator — the full pair list is never
    held in memory.  as_completed() ensures results flow out immediately
    without waiting for slower pairs to finish.
    """
    n = len(metas)
    n_pairs = n * (n - 1) // 2

    if verbose:
        print(
            f"\nDispatching {n_pairs} pairs across {workers} threads...",
            file=sys.stderr,
        )

    with ThreadPoolExecutor(max_workers=workers) as pool:
        # combinations() is lazy: pairs are generated on demand as the
        # dict comprehension iterates, not all at once.
        pending = {
            pool.submit(_compute_pair, a, b, alpha, beta): (a.path, b.path)
            for a, b in combinations(metas, 2)
        }
        completed = 0
        report_every = max(1, n_pairs // 20)
        for future in as_completed(pending):
            completed += 1
            if verbose and completed % report_every == 0:
                print(
                    f"  {completed}/{n_pairs}  ({100*completed//n_pairs}%)",
                    file=sys.stderr,
                )
            try:
                yield future.result()
            except Exception as e:
                a_path, b_path = pending[future]
                print(
                    f"  Warning: pair ({a_path.name}, {b_path.name}) failed: {e}",
                    file=sys.stderr,
                )


# ---------------------------------------------------------------------------
# Output
# ---------------------------------------------------------------------------

FIELDS = [
    "file_a", "file_b", "hybrid", "ncd_dict", "ncd_fingerprint",
    "entropy_global", "entropy_profile", "H_a", "H_b",
]


def stream_csv(results: Generator) -> None:
    """Stream results directly to stdout — never buffers all pairs."""
    writer = csv.DictWriter(sys.stdout, fieldnames=FIELDS)
    writer.writeheader()
    for row in results:
        writer.writerow(row)


def print_table(results: Generator, sort_by: str) -> None:
    """Buffer results for sorting, then print."""
    rows = sorted(results, key=lambda r: r[sort_by])
    if not rows:
        print("No pairs computed.")
        return
    col_a = max(len(r["file_a"]) for r in rows)
    col_b = max(len(r["file_b"]) for r in rows)
    header = (
        f"{'File A':<{col_a}}  {'File B':<{col_b}}  "
        f"{'Hybrid':>8}  {'NCD-Dict':>8}  {'NCD-FP':>8}  "
        f"{'H-Global':>8}  {'H-Profile':>9}  {'H(a)':>5}  {'H(b)':>5}"
    )
    sep = "-" * len(header)
    print(sep)
    print(header)
    print(sep)
    for r in rows:
        print(
            f"{r['file_a']:<{col_a}}  {r['file_b']:<{col_b}}  "
            f"{r['hybrid']:>8.4f}  {r['ncd_dict']:>8.4f}  {r['ncd_fingerprint']:>8.4f}  "
            f"{r['entropy_global']:>8.4f}  {r['entropy_profile']:>9.4f}  "
            f"{r['H_a']:>5.2f}  {r['H_b']:>5.2f}"
        )
    print(sep)


# ---------------------------------------------------------------------------
# File discovery
# ---------------------------------------------------------------------------

def collect_files(directory: Path, extensions, max_size: int) -> list:
    files = []
    for p in sorted(directory.rglob("*")):
        if not p.is_file():
            continue
        if extensions and p.suffix.lower() not in extensions:
            continue
        try:
            size = p.stat().st_size
        except OSError:
            continue
        if size == 0:
            continue
        if size > max_size:
            print(f"  Skipping {p.name} ({size:,}B > max {max_size:,}B)", file=sys.stderr)
            continue
        files.append(p)
    return files


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------

def main() -> None:
    parser = argparse.ArgumentParser(
        prog="filesim",
        description="Recursive file similarity: hybrid NCD + entropy distance.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=f"""
Concurrency note: zstd releases the GIL, so ThreadPoolExecutor provides true
parallel compression.  ProcessPoolExecutor is ~4x slower due to pickle
overhead.  Default workers: {DEFAULT_WORKERS} (cpu_count - 1).

Examples:
  filesim /path/to/dir
  filesim /path/to/dir --ext .bin .exe --sort ncd_dict
  filesim /path/to/dir --format csv > results.csv
  filesim /path/to/dir --workers 8 --verbose
        """,
    )
    parser.add_argument("directory", type=Path)
    parser.add_argument("--ext", nargs="+", metavar="EXT")
    parser.add_argument("--max-size", type=int, default=10 * 1024 * 1024, metavar="BYTES",
                        help="Skip files larger than N bytes (default: 10MB)")
    parser.add_argument("--alpha", type=float, default=0.5,
                        help="Weight for dictionary NCD (default: 0.5)")
    parser.add_argument("--beta", type=float, default=0.25,
                        help="Weight for fingerprint NCD (default: 0.25)")
    parser.add_argument("--workers", type=int, default=DEFAULT_WORKERS, metavar="N",
                        help=f"Thread pool size (default: {DEFAULT_WORKERS})")
    parser.add_argument("--sort", default="hybrid",
                        choices=["hybrid", "ncd_dict", "ncd_fingerprint",
                                 "entropy_global", "entropy_profile"])
    parser.add_argument("--format", default="table", choices=["table", "csv"])
    parser.add_argument("--verbose", action="store_true")

    args = parser.parse_args()

    if not args.directory.is_dir():
        print(f"Error: {args.directory} is not a directory", file=sys.stderr)
        sys.exit(1)
    if args.alpha + args.beta > 1.0:
        print("Error: alpha + beta must be <= 1.0", file=sys.stderr)
        sys.exit(1)
    if args.workers < 1:
        print("Error: --workers must be >= 1", file=sys.stderr)
        sys.exit(1)

    extensions = (
        [e if e.startswith(".") else f".{e}" for e in args.ext]
        if args.ext else None
    )

    if args.verbose:
        print(f"Scanning {args.directory}...", file=sys.stderr)

    files = collect_files(args.directory, extensions, args.max_size)

    if len(files) < 2:
        print(f"Need at least 2 files; found {len(files)}.", file=sys.stderr)
        sys.exit(1)

    if args.verbose:
        print(
            f"Found {len(files)} files."
            f" Preprocessing with {args.workers} threads...",
            file=sys.stderr,
        )

    metas = preprocess_files(files, args.workers, args.verbose)

    if len(metas) < 2:
        print("Too many files failed to load.", file=sys.stderr)
        sys.exit(1)

    results = pair_results(metas, args.alpha, args.beta, args.workers, args.verbose)

    if args.format == "csv":
        stream_csv(results)
    else:
        print_table(results, args.sort)


if __name__ == "__main__":
    main()

File: filesim/filesim.go
// Package filesim provides recursive file similarity analysis using a hybrid
// metric that blends Normalized Compression Distance (NCD) with entropy-based
// distances.
//
// It is a Go port of the original Python filesim tool. The four components of
// the hybrid score are:
//
//   - ncd_dict        : NCD where a zstd dictionary trained on one file is used
//     to compress the other (falls back to concatenation NCD
//     when a file is too small to train a dictionary).
//   - ncd_fingerprint : NCD over the packed per-block entropy "fingerprint",
//     which captures class-level similarity (e.g. two
//     encrypted blobs both look flat at ~8 bits/byte).
//   - entropy_global  : |H(a) - H(b)| / 8.
//   - entropy_profile : mean absolute difference of per-block entropy.
//
// # Concurrency model
//
// klauspost/compress releases nothing to a GIL — Go has none — and its
// Encoder.EncodeAll is safe for concurrent use. We therefore build one zstd
// encoder per file (with its trained dictionary baked in) exactly once during
// preprocessing and share it across every pair goroutine. This avoids the
// O(N^2) dictionary reconstruction the reference Python implementation paid.
//
// # Memory model
//
// After preprocessing we keep only derived metadata per file, not raw bytes.
// Pair computation re-reads files from disk on demand; for small datasets the
// OS page cache makes this nearly free, and for large datasets it trades memory
// for I/O — the correct tradeoff when N is large.
package filesim

import (
	"encoding/binary"
	"fmt"
	"io/fs"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
	"sync"

	"github.com/klauspost/compress/zstd"
)

// Tuning constants (mirrors the Python implementation).
const (
	ZstdLevel        = 10
	MinDictSamples   = 5
	MinSampleSize    = 8
	EntropyBlockSize = 256
)

// DefaultWorkers is cpu_count-1 (at least 1), matching the Python default.
func DefaultWorkers() int {
	if n := runtime.NumCPU() - 1; n > 0 {
		return n
	}
	return 1
}

// plainEnc is a shared, dictionary-less encoder. EncodeAll is concurrency-safe,
// so a single instance serves every goroutine.
var plainEnc *zstd.Encoder

func init() {
	enc, err := zstd.NewWriter(nil,
		zstd.WithEncoderLevel(zstd.EncoderLevelFromZstd(ZstdLevel)))
	if err != nil {
		panic(fmt.Sprintf("filesim: cannot create zstd encoder: %v", err))
	}
	plainEnc = enc
}

// FileMeta holds the derived metadata for one file. Raw bytes are released after
// preprocessing; only this survives into the pair phase.
type FileMeta struct {
	Path    string
	Size    int
	cx      int           // compressed size with no dictionary
	h       float64       // global Shannon entropy [0, 8]
	profile []float64     // per-block entropy values
	fp      []byte        // entropy fingerprint (packed float32s)
	dictEnc *zstd.Encoder // encoder carrying this file's trained dict; nil if none
}

// Result is the hybrid similarity record for one unordered pair.
type Result struct {
	FileA          string
	FileB          string
	Hybrid         float64
	NCDDict        float64
	NCDFingerprint float64
	EntropyGlobal  float64
	EntropyProfile float64
	HA             float64
	HB             float64
}

// ---------------------------------------------------------------------------
// Entropy utilities
// ---------------------------------------------------------------------------

func shannonEntropy(data []byte) float64 {
	if len(data) == 0 {
		return 0.0
	}
	var counts [256]int
	for _, b := range data {
		counts[b]++
	}
	n := float64(len(data))
	h := 0.0
	for _, c := range counts {
		if c == 0 {
			continue
		}
		p := float64(c) / n
		h -= p * math.Log2(p)
	}
	return h
}

func entropyProfile(data []byte) []float64 {
	var result []float64
	for i := 0; i < len(data); i += EntropyBlockSize {
		end := min(i+EntropyBlockSize, len(data))
		blk := data[i:end]
		if len(blk) >= MinSampleSize {
			result = append(result, shannonEntropy(blk))
		}
	}
	if len(result) == 0 {
		return []float64{shannonEntropy(data)}
	}
	return result
}

func entropyFingerprint(profile []float64) []byte {
	buf := make([]byte, 4*len(profile))
	for i, v := range profile {
		binary.LittleEndian.PutUint32(buf[i*4:], math.Float32bits(float32(v)))
	}
	return buf
}

func profileDistance(p1, p2 []float64) float64 {
	n := min(len(p1), len(p2))
	if n == 0 {
		return 1.0
	}
	sum := 0.0
	for i := range n {
		sum += math.Abs(p1[i] - p2[i])
	}
	return sum / (float64(n) * 8.0)
}

// ---------------------------------------------------------------------------
// Compression utilities
// ---------------------------------------------------------------------------

func compressSize(enc *zstd.Encoder, data []byte) int {
	return len(enc.EncodeAll(data, nil))
}

// buildDictEncoder trains a zstd dictionary on data and returns an encoder that
// uses it. Returns nil when data is too small to yield enough samples or when
// training fails — callers then fall back to concatenation NCD.
func buildDictEncoder(data []byte) *zstd.Encoder {
	chunk := max(len(data)/10, MinSampleSize)
	var samples [][]byte
	for i := 0; i < len(data); i += chunk {
		end := min(i+chunk, len(data))
		if end-i >= MinSampleSize {
			// copy: BuildDict retains references to the sample slices.
			s := make([]byte, end-i)
			copy(s, data[i:end])
			samples = append(samples, s)
		}
	}
	if len(samples) < MinDictSamples {
		return nil
	}
	dictBytes, err := zstd.BuildDict(zstd.BuildDictOptions{
		Contents: samples,
		Level:    zstd.EncoderLevelFromZstd(ZstdLevel),
	})
	if err != nil {
		return nil
	}
	enc, err := zstd.NewWriter(nil,
		zstd.WithEncoderLevel(zstd.EncoderLevelFromZstd(ZstdLevel)),
		zstd.WithEncoderDict(dictBytes))
	if err != nil {
		return nil
	}
	return enc
}

// ---------------------------------------------------------------------------
// Preprocessing: file -> FileMeta (raw bytes released on return)
// ---------------------------------------------------------------------------

func preprocessFile(path string) (*FileMeta, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	profile := entropyProfile(data)
	return &FileMeta{
		Path:    path,
		Size:    len(data),
		cx:      compressSize(plainEnc, data),
		h:       shannonEntropy(data),
		profile: profile,
		fp:      entropyFingerprint(profile),
		dictEnc: buildDictEncoder(data),
	}, nil
}

// Preprocess reads and derives metadata for each path concurrently. Files that
// cannot be read are reported via warn (if non-nil) and skipped.
func Preprocess(paths []string, workers int, warn func(string)) []*FileMeta {
	if workers < 1 {
		workers = 1
	}
	jobs := make(chan string)
	var mu sync.Mutex
	var metas []*FileMeta
	var wg sync.WaitGroup

	for range workers {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for p := range jobs {
				meta, err := preprocessFile(p)
				if err != nil {
					if warn != nil {
						warn(fmt.Sprintf("cannot read %s: %v", p, err))
					}
					continue
				}
				mu.Lock()
				metas = append(metas, meta)
				mu.Unlock()
			}
		}()
	}
	for _, p := range paths {
		jobs <- p
	}
	close(jobs)
	wg.Wait()
	return metas
}

// ---------------------------------------------------------------------------
// Pair computation
// ---------------------------------------------------------------------------

func ncdFP(fpA, fpB []byte) float64 {
	ca := compressSize(plainEnc, fpA)
	cb := compressSize(plainEnc, fpB)
	cat := make([]byte, 0, len(fpA)+len(fpB))
	cat = append(append(cat, fpA...), fpB...)
	cab := compressSize(plainEnc, cat)
	denom := max(ca, cb)
	if denom == 0 {
		return 0.0
	}
	return float64(cab-min(ca, cb)) / float64(denom)
}

func clamp01(v float64) float64 {
	if v < 0 {
		return 0
	}
	if v > 1 {
		return 1
	}
	return v
}

func computePair(a, b *FileMeta, alpha, beta float64) (Result, error) {
	dataA, err := os.ReadFile(a.Path)
	if err != nil {
		return Result{}, err
	}
	dataB, err := os.ReadFile(b.Path)
	if err != nil {
		return Result{}, err
	}

	denom := max(a.cx, b.cx)

	var nd float64
	if a.dictEnc != nil || b.dictEnc != nil {
		// Conditional NCD: compress each file using a dictionary trained on the
		// other, then take the cheaper direction. Missing dict => plain encoder.
		var cBgivenA int
		if a.dictEnc != nil {
			cBgivenA = compressSize(a.dictEnc, dataB)
		} else {
			cBgivenA = compressSize(plainEnc, dataB)
		}
		var cAgivenB int
		if b.dictEnc != nil {
			cAgivenB = compressSize(b.dictEnc, dataA)
		} else {
			cAgivenB = compressSize(plainEnc, dataA)
		}
		if denom != 0 {
			nd = float64(min(cBgivenA, cAgivenB)) / float64(denom)
		}
	} else {
		cat := make([]byte, 0, len(dataA)+len(dataB))
		cat = append(append(cat, dataA...), dataB...)
		cxy := compressSize(plainEnc, cat)
		if denom != 0 {
			nd = float64(cxy-min(a.cx, b.cx)) / float64(denom)
		}
	}
	nd = clamp01(nd)

	nf := clamp01(ncdFP(a.fp, b.fp))
	eg := math.Abs(a.h-b.h) / 8.0
	ep := profileDistance(a.profile, b.profile)

	gamma := (1.0 - alpha - beta) / 2.0
	hybrid := alpha*nd + beta*nf + gamma*eg + gamma*ep

	return Result{
		FileA:          a.Path,
		FileB:          b.Path,
		Hybrid:         round6(hybrid),
		NCDDict:        round6(nd),
		NCDFingerprint: round6(nf),
		EntropyGlobal:  round6(eg),
		EntropyProfile: round6(ep),
		HA:             round3(a.h),
		HB:             round3(b.h),
	}, nil
}

func round6(v float64) float64 { return math.Round(v*1e6) / 1e6 }
func round3(v float64) float64 { return math.Round(v*1e3) / 1e3 }

// Pairs computes every unordered pair concurrently and streams results on the
// returned channel, which is closed when all pairs are done. Pairs are
// generated lazily, so the full O(N^2) list is never materialized. warn, if
// non-nil, receives per-pair failure messages.
func Pairs(metas []*FileMeta, alpha, beta float64, workers int, warn func(string)) <-chan Result {
	if workers < 1 {
		workers = 1
	}
	out := make(chan Result)

	type job struct{ a, b *FileMeta }
	jobs := make(chan job, workers*2)

	go func() {
		for i := range metas {
			for j := i + 1; j < len(metas); j++ {
				jobs <- job{metas[i], metas[j]}
			}
		}
		close(jobs)
	}()

	var wg sync.WaitGroup
	for range workers {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for jb := range jobs {
				res, err := computePair(jb.a, jb.b, alpha, beta)
				if err != nil {
					if warn != nil {
						warn(fmt.Sprintf("pair (%s, %s) failed: %v",
							filepath.Base(jb.a.Path), filepath.Base(jb.b.Path), err))
					}
					continue
				}
				out <- res
			}
		}()
	}
	go func() {
		wg.Wait()
		close(out)
	}()
	return out
}

// NumPairs returns n*(n-1)/2.
func NumPairs(n int) int { return n * (n - 1) / 2 }

// ---------------------------------------------------------------------------
// File discovery
// ---------------------------------------------------------------------------

// CollectFiles walks dir recursively, returning files that pass the extension
// and size filters in sorted order. exts entries should include the leading dot
// and be lowercase; an empty exts means "any extension". skip, if non-nil,
// receives messages about skipped files.
func CollectFiles(dir string, exts []string, maxSize int64, skip func(string)) ([]string, error) {
	extSet := make(map[string]bool, len(exts))
	for _, e := range exts {
		extSet[strings.ToLower(e)] = true
	}

	var files []string
	err := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil // ignore unreadable entries, keep walking
		}
		if d.IsDir() {
			return nil
		}
		if len(extSet) > 0 && !extSet[strings.ToLower(filepath.Ext(path))] {
			return nil
		}
		info, err := d.Info()
		if err != nil {
			return nil
		}
		size := info.Size()
		if size == 0 {
			return nil
		}
		if size > maxSize {
			if skip != nil {
				skip(fmt.Sprintf("skipping %s (%d B > max %d B)",
					filepath.Base(path), size, maxSize))
			}
			return nil
		}
		files = append(files, path)
		return nil
	})
	if err != nil {
		return nil, err
	}
	sort.Strings(files)
	return files, nil
}

// SortKey returns the comparable value of r for the given field name, used to
// sort table output. Unknown fields fall back to Hybrid.
func SortKey(r Result, by string) float64 {
	switch by {
	case "ncd_dict":
		return r.NCDDict
	case "ncd_fingerprint":
		return r.NCDFingerprint
	case "entropy_global":
		return r.EntropyGlobal
	case "entropy_profile":
		return r.EntropyProfile
	default:
		return r.Hybrid
	}
}
